src/main/resources/lib/guillotine/index.js

```diff
@@ -65,10 +67,83 @@ function createContext(options) {
     }
     context.options.applications = context.options.applications || [app.name];
     context.options.allowPaths = context.options.allowPaths || [];
+    context.options.subscriptionEventTypes = context.options.subscriptionEventTypes || ['node.*'];
 
     return context;
 }
 
+//──────────────────────────────────────────────────────────────────────────────
+// GraphQL WS Protocol
+//──────────────────────────────────────────────────────────────────────────────
+const graphQlSubscribers = {};
+
+function cancelSubscription(sessionId) {
+    Java.synchronized(() => {
+        const subscriber = graphQlSubscribers[sessionId];
+        if (subscriber) {
+            delete graphQlSubscribers[sessionId];
+            subscriber.cancelSubscription();
+        }
+    }, graphQlSubscribers)();
+}
+
+function handleStartMessage(schema, sessionId, message) {
+    const graphQLOperationId = message.id;
+    const payload = message.payload;
+
+    try {
+        const result = graphQlLib.execute(schema, payload.query, payload.variables);
+
+        if (result.data instanceof com.enonic.lib.graphql.rx.Publisher) {
+            const subscriber = graphQlRxLib.createSubscriber({
+                onNext: (result) => {
+                    webSocketLib.send(sessionId, JSON.stringify({
+                        type: 'data',
+                        id: graphQLOperationId,
+                        payload: result
+                    }));
+                }
+            });
+            Java.synchronized(() => graphQlSubscribers[sessionId] = subscriber, graphQlSubscribers)();
+            result.data.subscribe(subscriber);
+        }
+    } catch (e) {
+        log.error('Error while handling Start GraphQL-WS message', e);
+        throw e;
+    }
+}
+
+function initWebSockets(schema) {
+    return function (event) {
+        switch (event.type) {
+        case 'close':
+            cancelSubscription(event.session.id);
+            break;
+        case 'message':
+            var message = JSON.parse(event.message);
+            switch (message.type) {
```

Comment:
switch/case in switch/case is hard to read

Reply:
done